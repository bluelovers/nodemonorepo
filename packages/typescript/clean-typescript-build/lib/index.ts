import {partition} from 'ramda'
import {unlink, lstat, existsSync} from 'fs-extra'
import {traverse} from 'fs-tree-utils'

import {
  Options,
  TargetList,
  Clean
} from './types'

import {
  DEFAULT_DEEP_FUNC,
  DEFAULT_SOURCE_DETECTOR,
  DEFAULT_TARGET_SPECIFIER
} from './contants'

export * from './types'
export * from './contants'

/**
 * Delete all compiled products corresponding to source files inside `root`
 * @param root Directory that contains all source files
 * @param options Options
 * @param options.deep A function that tells when to go deeper
 * @param options.isSource A function that tells which file is a source file
 * @param options.listTargets A function that lists compiled products corresponding to a source file
 * @returns An object
 *   * Property `targets`: An aggregation of `options.listTargets`'s return values
 *   * Property `reports`: An array of reports which tell whether or not deletion succeed
 *   * Property `success`: An array of paths to files which are successfully deleted
 *   * Property `failure`: An array of paths to files which are failed to be deleted
 */
export async function clean (root: string, options?: Options): Promise<Clean.Result> {
  const del = (file: string) => unlink(file).then(onUnlinkSuccess, onUnlinkFailure)
  const r2f = (reports: Clean.Result.ReportList) => reports.map(x => x.file)
  const onUnlinkSuccess = () => ({success: true as true})
  const onUnlinkFailure = (error: any) => ({success: false as false, error})
  const targets = await listAllTargets(root, options)
  const reports = await Promise.all(targets.map(async file => ({file, deletion: await del(file)})))
  const [success, failure] = partition(x => x.deletion.success, reports)
  return {targets, reports, success: r2f(success), failure: r2f(failure)}
}

/**
 * List compiled products corresponding to source files inside `root`
 * @param root Directory that contains all source files
 * @param options Options
 * @param options.deep A function that tells when to go deeper
 * @param options.isSource A function that tells which file is a source file
 * @param options.listTargets A function that lists compiled products corresponding to a source file
 * @returns A promise that resolves a list of target file paths
 */
export async function listAllTargets (
  root: string,
  options: Options = {}
): Promise<TargetList> {
  const {
    deep = DEFAULT_DEEP_FUNC,
    isSource = DEFAULT_SOURCE_DETECTOR,
    listTargets = DEFAULT_TARGET_SPECIFIER
  } = options

  const existed = (await traverse(root, {deep, stat: lstat}))
    .filter(isSource)
    .map(listTargets)
    .reduce((prev, current) => [...prev, ...current], [])
    .filter(item => existsSync(item))

  const files = (await Promise.all(
    existed.map(async item => ({
      item,
      stats: await lstat(item)
    }))
  ))
    .filter(x => x.stats.isFile())
    .map(x => x.item)

  return files
}
