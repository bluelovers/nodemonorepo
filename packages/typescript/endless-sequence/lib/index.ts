import EndlessIterable from './iterable'
import * as applications from './applications'
export {EndlessIterable, applications}
