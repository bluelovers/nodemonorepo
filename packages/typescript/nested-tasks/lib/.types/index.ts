import * as childProcess from 'child_process'
import shCmd from 'sh-or-cmd'
import {Traverse} from 'fs-tree-utils'
import TraversalDeepFuncParam = Traverse.Options.DeepFunc.Param

import {
  TaskParam,
  TaskSetManifest,
  DependencyList
} from '../../api'

export namespace Manifest {
  export interface Descriptor extends TraversalDeepFuncParam {
    readonly type: Type
  }

  export type Type = 'module' | 'yaml'
}

export class Task {
  static readonly shell: string = shCmd

  //@ts-ignore
  readonly before: DependencyList

  //@ts-ignore
  readonly parallel: DependencyList

  //@ts-ignore
  readonly after: DependencyList

  //@ts-ignore
  readonly command: Task.Command

  //@ts-ignore
  readonly execute: Task.Executor

  constructor (info: TaskParam) {
    type TpsShell = TaskParam.Shortcut.Shell
    type TpsFunction = TaskParam.Shortcut.Function

    switch (typeof info) {
      case 'string':
        return new this.Task({type: 'shell', cmd: info as TpsShell})

      case 'function':
        return new this.Task({type: 'function', execute: info as TpsFunction})

      case 'object':
        break // main branch: below

      default:
        throw new TypeError(`Invalid type of info: ${JSON.stringify(info)}`)
    }

    if (Array.isArray(info)) {
      const [program, ...argv] = info
      return new Task({type: 'spawn', program, argv})
    }

    {
      this.before = createDependencyList('before')
      this.parallel = createDependencyList('parallel')
      this.after = createDependencyList('after')

      function createDependencyList (name: keyof TaskParam.utils.Base): DependencyList {
        const array = (info as TaskParam.utils.Base)[name] || []

        if (!Array.isArray(array)) {
          throw new TypeError(`Property '${name}' is not an array: ${JSON.stringify(array)}`)
        }

        const result = Array<DependencyList.TaskName>()

        for (const item of array) {
          if (Array.isArray(item)) {
            result.push(item)
            continue
          }

          if (typeof item === 'string') {
            result.push(item.split(/\s|\./))
            continue
          }

          throw new TypeError(`Invalid type of an item: ${JSON.stringify(item)}`)
        }

        return result
      }
    }

    const infoAsObject = info as TaskParam.Detailed | TaskParam.Keyword

    if ('type' in infoAsObject) {
      type Shell = TaskParam.Detailed.Shell
      type Spawn = TaskParam.Detailed.Spawn
      type Function = TaskParam.Detailed.Function

      const {type} = infoAsObject

      if (typeof type !== 'string') {
        throw new TypeError(`Property 'type' is not a string: ${JSON.stringify(type)}`)
      }

      switch (type) {
        case 'shell': {
          const {
            shell = this.Task.shell,
            cmd,
            options,
            oncreate = () => {}
          } = infoAsObject as Shell

          if (typeof shell !== 'string') {
            throw new TypeError(`Property 'shell' is not a string: ${JSON.stringify(shell)}`)
          }

          if (typeof cmd !== 'string') {
            throw new TypeError(`Property 'cmd' is not a string: ${JSON.stringify(cmd)}`)
          }

          if (typeof oncreate !== 'function') {
            throw new TypeError(`Property 'oncreate' is not a function: ${JSON.stringify(oncreate)}`)
          }

          this.command = {type, shell, cmd, options, oncreate}

          this.execute = createChildProcessExecutor(
            () => childProcess.spawn(shell, [], {...options}),
            x => {
              x.stdin.end(shell + '\n')
              oncreate(x)
            }
          )

          return
        }

        case 'spawn': {
          const {
            program,
            argv,
            options,
            oncreate = () => {}
          } = infoAsObject as Spawn

          if (typeof program !== 'string') {
            throw new TypeError(`Property 'program' is not a string: ${JSON.stringify(program)}`)
          }

          if (!Array.isArray(argv)) {
            throw new TypeError(`Property 'argv' is not an array: ${JSON.stringify(argv)}`)
          }

          if (typeof oncreate !== 'function') {
            throw new TypeError(`Property 'oncreate' is not a function: ${JSON.stringify(oncreate)}`)
          }

          this.command = {type, program, argv, options, oncreate}

          this.execute = createChildProcessExecutor(
            () => childProcess.spawn(program, argv, {...options}),
            oncreate
          )

          return
        }

        case 'function': {
          const {execute} = infoAsObject as Function

          if (typeof execute !== 'function') {
            throw new TypeError(`Property 'execute' is not a function: ${JSON.stringify(execute)}`)
          }

          this.command = {type, execute}
          this.execute = async () => void await middle()

          async function middle () {
            return execute()
          }

          return
        }

        //@ts-ignore Unreachable
        throw new TypeError(`Invalid type of info: ${JSON.stringify(infoAsObject)} (${typeof infoAsObject})`)

        function createChildProcessExecutor (
          fn: () => childProcess.ChildProcess,
          oncreate: TaskParam.Detailed.utils.ChildProcess.CreationHandler
        ): Task.Executor {
          return () => {
            const cp = fn()
            oncreate(cp)
            return createChildProcessPromise(cp)
          }
        }

        function createChildProcessPromise (cp: childProcess.ChildProcess): Promise<void> {
          return new Promise((resolve, reject) => {
            cp.on('error', error => reject(error))

            cp.on(
              'close',
              (status, signal) => status
                ? reject(new Task.ChildProcessError(status, signal))
                : resolve()
            )
          })
        }
      }
    }

    if ('cmd' in infoAsObject) {
      return new this.Task({type: 'shell', ...infoAsObject})
    }

    if ('spawn' in infoAsObject) {
      const {spawn, ...rest} = infoAsObject

      if (!Array.isArray(spawn)) {
        throw new TypeError(`Property 'spawn' is not an array: ${JSON.stringify(spawn)}`)
      }

      const [program, ...argv] = spawn
      return new this.Task({type: 'spawn', program, argv, ...rest})
    }

    if ('execute' in infoAsObject) {
      return new this.Task({type: 'function', ...infoAsObject})
    }

    if ('program' in infoAsObject) {
      const {program, argv = [], ...rest} = infoAsObject

      if (typeof program !== 'string') {
        throw new TypeError(`Property 'program' is not a string: ${JSON.stringify(program)}`)
      }

      if (!Array.isArray(argv)) {
        throw new TypeError(`Property 'argv' is not an array: ${JSON.stringify(argv)}`)
      }

      return new this.Task({type: 'spawn', program, argv, ...rest})
    }

    throw new TypeError(`Invalid set of properties: ${JSON.stringify(Object.keys(info))}`)
  }

  get Task () {
    return this.constructor as typeof Task
  }
}

export namespace Task {
  export type Command =
    Command.Shell |
    Command.Spawn |
    Command.Function

  export namespace Command {
    import api = TaskParam.Detailed

    export interface Shell extends api.Shell {
      readonly type: 'shell'
      readonly shell: string
      readonly cmd: string
    }

    export interface Spawn extends api.Spawn {
      readonly type: 'spawn'
      readonly program: string
      readonly argv: ReadonlyArray<string>
    }

    export interface Function extends api.Function {
      readonly type: 'function'
      readonly execute: () => void
    }
  }

  export type Executor = () => Promise<void>

  export class ChildProcessError extends Error {
    readonly status: number
    readonly signal: string | null

    constructor (status: number, signal: string | null) {
      super(`Child process exit with code ${status} (signal: ${signal})`)
      this.status = status
      this.signal = signal
    }
  }
}

export class TaskSet {
  readonly manifest: object
  readonly tasks: TaskDict
  readonly subtasks: SubtaskDict

  constructor (manifest: TaskSetManifest) {
    const tasks: {[_: string]: Task} = {}
    const subtasks: {[_: string]: TaskSet} = {}

    this.manifest = manifest
    this.tasks = tasks
    this.subtasks = subtasks

    for (const [name, info] of Object.entries(manifest)) {
      switch (classifyPropertyName(name)) {
        case classifyPropertyName.Result.Invalid:
          throw new Error(`Invalid name: ${name}`)
        case classifyPropertyName.Result.Task:
          tasks[name] = new Task(info as TaskParam)
          break
        case classifyPropertyName.Result.Subtask:
          subtasks[name.slice(1)] = new TaskSet(info as TaskSetManifest)
          break
      }
    }
  }
}

export interface TaskDict {
  readonly [name: string]: Task | void
}

export interface SubtaskDict {
  readonly [name: string]: TaskSet | void
}

export function classifyPropertyName (name: string): classifyPropertyName.Result {
  const {Result} = classifyPropertyName
  if (/^[a-z-_]+$/i.test(name)) return Result.Task
  if (/^@[a-z-_]+$/i.test(name)) return Result.Subtask
  return Result.Invalid
}

export namespace classifyPropertyName {
  export enum Result {
    Invalid,
    Task,
    Subtask
  }
}
