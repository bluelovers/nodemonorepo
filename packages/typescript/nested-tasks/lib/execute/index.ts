import {Task, TaskSet} from '../.types'
import loadManifestFiles from '../load-manifest-files'

/**
 *
 * @param dirname Directory that contains all task file
 * @param taskpath Dot separated task names
 * @param options Options to pass to `loadManifestFiles`
 */
export async function execute (
  dirname: string,
  taskpath: string,
  options?: execute.Options
) {
  return execute.fromList(
    await loadManifestFiles(dirname, options),
    taskpath.split('.')
  )
}

export namespace execute {
  export async function fromList (
    tasklist: ReadonlyArray<loadManifestFiles.Result.Item>,
    taskpath: ReadonlyArray<string>
  ) {}

  export type Options = loadManifestFiles.Options
}

export default execute
