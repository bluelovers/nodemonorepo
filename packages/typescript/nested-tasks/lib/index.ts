import listManifestFiles from './list-manifest-files'
export {listManifestFiles}
