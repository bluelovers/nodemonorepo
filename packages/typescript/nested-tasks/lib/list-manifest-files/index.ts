import * as path from 'path'
import * as fsTreeUtils from 'fs-tree-utils'

import {
  DeepFunc as PrvDeepFunc,
  TraversalDeepFuncParam
} from 'fs-tree-utils/lib/traverse'

/**
 * @param dirname Directory that contains all task manifest descriptors
 * @param options Specify deep function and file chooser
 * @returns A list of task manifests
 */
export async function listManifestFiles (
  dirname: string,
  options: listManifestFiles.Options = {}
): listManifestFiles.Result {
  const {
    deep = listManifestFiles.DEFAULT_DEEP_FUNCTION,
    choose = listManifestFiles.DEFAULT_FILE_CHOOSER
  } = options

  const list = await fsTreeUtils.traverse(dirname, deep)
  const result = Array<listManifestFiles.ManifestDescriptor>()

  for (const item of list) {
    const type = choose(item)
    if (!type) continue
    result.push({...item, type})
  }

  return result
}

export namespace listManifestFiles {
  export const DEFAULT_DEEP_FUNCTION: DeepFunc = x => x.item !== 'node_modules'

  export const DEFAULT_FILE_CHOOSER: FileChooser = x => {
    switch (path.extname(x.item)) {
      case '':
      case '.js':
        return 'module'
      case '.json':
      case '.yaml':
      case '.yml':
        return 'yaml'
    }

    return null
  }

  export interface Options {
    readonly deep?: DeepFunc
    readonly choose?: FileChooser
  }

  export type Result = Promise<ReadonlyArray<ManifestDescriptor>>
  export type DeepFunc = PrvDeepFunc
  export type FileChooser = (param: FileChooser.Param) => ManifestType | null | void

  export namespace FileChooser {
    export type Param = TraversalDeepFuncParam
  }

  export interface ManifestDescriptor extends TraversalDeepFuncParam {
    readonly type: ManifestType
  }

  export type ManifestType = 'module' | 'yaml'
}

export default listManifestFiles
