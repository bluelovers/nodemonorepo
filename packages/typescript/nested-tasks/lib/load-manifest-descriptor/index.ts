import * as childProcess from 'child_process'
import yaml from 'monorepo-shared-yaml'
import {Manifest} from '../.types'

/**
 * @param descriptor A manifest descriptor
 * @returns A task set
 */
export async function loadManifestDescriptor (
  descriptor: Manifest.Descriptor
): Promise<loadManifestDescriptor.TaskSet> {
  const manifest = await loadObject(descriptor)

  if (typeof manifest !== 'object') {
    throw new TypeError(`Manifest is not an object: ${manifest}`)
  }

  if (manifest instanceof Array) {
    throw new TypeError(`Manifest is an array: ${JSON.stringify(manifest)}`)
  }

  if (!manifest) {
    throw new TypeError('Manifest is null')
  }

  return new loadManifestDescriptor.TaskSet(manifest)
}

export namespace loadManifestDescriptor {
  export class Task {
    static readonly shell: string = 'sh'

    //@ts-ignore
    readonly before: Task.DependencyList

    //@ts-ignore
    readonly after: Task.DependencyList

    //@ts-ignore
    readonly command: Task.Command

    //@ts-ignore
    readonly execute: Task.Executor

    constructor (info: any) {
      switch (typeof info) {
        case 'string':
          return new this.Task({type: 'shell', cmd: info})

        case 'function':
          return new this.Task({type: 'function', execute: info})

        case 'object':
          break // main branch: below

        default:
          throw new TypeError(`Invalid type of info: ${JSON.stringify(info)}`)
      }

      if (info instanceof Array) {
        const [program, ...argv] = info
        return new Task({type: 'spawn', program, argv})
      }

      {
        this.before = createDependencyList('before')
        this.after = createDependencyList('after')

        function createDependencyList (name: string): Task.DependencyList {
          const array = info[name] || []

          if (!Array.isArray(array)) {
            throw new TypeError(`Property '${name}' is not an array: ${JSON.stringify(array)}`)
          }

          const result = Array<Task.DependencyList.TaskName>()

          for (const item of array) {
            if (Array.isArray(item)) {
              result.push(item)
              continue
            }

            if (typeof item === 'string') {
              result.push(item.split(/\s|\./))
              continue
            }

            throw new TypeError(`Invalid type of an item: ${JSON.stringify(item)}`)
          }

          return result
        }
      }

      if ('type' in info) {
        const {type} = info

        if (typeof type !== 'string') {
          throw new TypeError(`Property 'type' is not a string: ${JSON.stringify(type)}`)
        }

        switch (type) {
          case 'shell': {
            const {
              shell = this.Task.shell,
              cmd,
              options,
              oncreate = () => {}
            } = info

            if (typeof shell !== 'string') {
              throw new TypeError(`Property 'shell' is not a string: ${JSON.stringify(shell)}`)
            }

            if (typeof cmd !== 'string') {
              throw new TypeError(`Property 'cmd' is not a string: ${JSON.stringify(cmd)}`)
            }

            if (typeof oncreate !== 'function') {
              throw new TypeError(`Property 'oncreate' is not a function: ${JSON.stringify(oncreate)}`)
            }

            const command: Task.Command.Shell = {type, shell, cmd, options, oncreate}
            this.command = command

            this.execute = createChildProcessExecutor(
              () => childProcess.spawn(shell, {...options}),
              oncreate
            )

            return
          }

          case 'spawn': {
            const {
              program,
              argv,
              options,
              oncreate = () => {}
            } = info

            if (typeof program !== 'string') {
              throw new TypeError(`Property 'program' is not a string: ${JSON.stringify(program)}`)
            }

            if (!Array.isArray(argv)) {
              throw new TypeError(`Property 'argv' is not an array: ${JSON.stringify(argv)}`)
            }

            if (typeof oncreate !== 'function') {
              throw new TypeError(`Property 'oncreate' is not a function: ${JSON.stringify(oncreate)}`)
            }

            const command: Task.Command.Spawn = {type, program, argv, options, oncreate}
            this.command = command

            this.execute = createChildProcessExecutor(
              () => childProcess.spawn(program, argv, {...options}),
              oncreate
            )

            return
          }

          case 'function': {
            const {execute} = info

            if (typeof execute !== 'function') {
              throw new TypeError(`Property 'execute' is not a function: ${JSON.stringify(execute)}`)
            }

            this.command = {type, execute}
            this.execute = async () => void await middle()

            async function middle () {
              return execute()
            }

            return
          }
        }

        throw new TypeError(`Invalid type of info: ${JSON.stringify(info)} (${typeof info})`)

        function createChildProcessExecutor (
          fn: () => childProcess.ChildProcess,
          oncreate: Task.Command.utils.ChildProcess.CreationHandler
        ): Task.Executor {
          return () => {
            const cp = fn()
            oncreate(cp)
            return createChildProcessPromise(cp)
          }
        }

        function createChildProcessPromise (cp: childProcess.ChildProcess): Promise<void> {
          return new Promise((resolve, reject) => {
            cp.on('error', error => reject(error))

            cp.on(
              'close',
              (status, signal) => status
                ? reject(new Task.ChildProcessError(status, signal))
                : resolve()
            )
          })
        }
      }

      if ('cmd' in info) {
        return new this.Task({type: 'shell', ...info})
      }

      if ('spawn' in info) {
        const {spawn, ...rest} = info

        if (!Array.isArray(spawn)) {
          throw new TypeError(`Property 'spawn' is not an array: ${JSON.stringify(spawn)}`)
        }

        const [program, ...argv] = spawn
        return new this.Task({type: 'spawn', program, argv, ...rest})
      }

      if ('execute' in info) {
        return new this.Task({type: 'function', ...info})
      }

      throw new TypeError(`Invalid set of properties: ${JSON.stringify(Object.keys(info))}`)
    }

    get Task () {
      return this.constructor as typeof Task
    }
  }

  export namespace Task {
    export type DependencyList = ReadonlyArray<DependencyList.TaskName>

    export namespace DependencyList {
      export type TaskName = ReadonlyArray<string>
    }

    export type Command =
      Command.Shell |
      Command.Spawn |
      Command.Function

    export namespace Command {
      export interface Shell extends utils.ChildProcess {
        readonly type: 'shell'
        readonly shell: string
        readonly cmd: string
      }

      export interface Spawn extends utils.ChildProcess {
        readonly type: 'spawn'
        readonly program: string
        readonly argv: ReadonlyArray<string>
      }

      export interface Function {
        readonly type: 'function'
        readonly execute: () => void
      }

      export namespace utils {
        export interface ChildProcess {
          readonly options?: childProcess.SpawnOptions
          readonly oncreate?: ChildProcess.CreationHandler
        }

        export namespace ChildProcess {
          export type CreationHandler = (x: childProcess.ChildProcess) => void
        }
      }
    }

    export type Executor = () => Promise<void>

    export class ChildProcessError extends Error {
      readonly status: number
      readonly signal: string | null

      constructor (status: number, signal: string | null) {
        super(`Child process exit with code ${status} (signal: ${signal})`)
        this.status = status
        this.signal = signal
      }
    }
  }

  export class TaskSet {
    readonly manifest: object
    readonly tasks: TaskDict
    readonly subtasks: SubtaskDict

    constructor (manifest: object) {
      const tasks: {[_: string]: Task} = {}
      const subtasks: {[_: string]: TaskSet} = {}

      this.manifest = manifest
      this.tasks = tasks
      this.subtasks = subtasks

      for (const [name, info] of Object.entries(manifest)) {
        switch (classifyPropertyName(name)) {
          case classifyPropertyName.Result.Invalid:
            throw new Error(`Invalid name: ${name}`)
          case classifyPropertyName.Result.Task:
            tasks[name] = new Task(info)
            break
          case classifyPropertyName.Result.Subtask:
            subtasks[name] = new TaskSet(info)
            break
        }
      }
    }
  }

  export interface TaskDict {
    readonly [name: string]: Task | void
  }

  export interface SubtaskDict {
    readonly [name: string]: TaskSet | void
  }

  export function classifyPropertyName (name: string): classifyPropertyName.Result {
    const {Result} = classifyPropertyName
    if (/^[a-z-_]+$/i.test(name)) return Result.Task
    if (/^@[a-z-_]+$/i.test(name)) return Result.Subtask
    return Result.Invalid
  }

  export namespace classifyPropertyName {
    export enum Result {
      Invalid,
      Task,
      Subtask
    }
  }
}

async function loadObject (descriptor: Manifest.Descriptor): Promise<any> {
  const {type, path} = descriptor

  switch (type) {
    case 'module':
      return require(path)
    case 'yaml':
      return await yaml.loadFile(path)
  }
}

export default loadManifestDescriptor
