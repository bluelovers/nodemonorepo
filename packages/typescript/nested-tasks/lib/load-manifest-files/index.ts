import listManifestFiles from '../list-manifest-files'
import loadManifestDescriptor from '../load-manifest-descriptor'
import {Manifest, TaskSet} from '../.types'

/**
 * @param dirname Directory that contains all task manifest descriptors
 * @param options Specify deep function and file chooser
 * @returns A list of task collections
 */
export async function loadManifestFiles (
  dirname: string,
  options?: loadManifestFiles.Options
): loadManifestFiles.Result {
  const list = await listManifestFiles(dirname, options)
  return loadManifestFiles.fromList(list)
}

export namespace loadManifestFiles {
  /**
   * @param list A list of task manifests
   * @returns A list of task collections
   */
  export async function fromList (list: List): Result {
    return Promise.all(
      list.map(async descriptor => ({
        descriptor,
        tasks: await loadManifestDescriptor(descriptor)
      }))
    )
  }

  export type List = ReadonlyArray<Manifest.Descriptor>

  export type Options = listManifestFiles.Options

  export type Result = Promise<ReadonlyArray<Result.Item>>

  export namespace Result {
    export interface Item {
      readonly descriptor: Manifest.Descriptor
      readonly tasks: TaskSet
    }
  }
}

export default loadManifestFiles
