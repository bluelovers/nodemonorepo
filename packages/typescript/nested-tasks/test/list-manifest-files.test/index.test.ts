import {listManifestFiles} from '../../index'
import createVirtualEnvironment from '../.lib/virtual-env'
import {chooseTypeScript, ignoreCustomDeepFunc} from '../.lib/utils'

describe('without javascript files', () => {
  const {apply} = createVirtualEnvironment()

  it('without options', apply(async () => {
    const sample = await getSample('root')
    expect(sample).toMatchSnapshot()
  }))

  it('with custom chooser', apply(async () => {
    const sample = await getSample('root', {
      choose: chooseTypeScript
    })

    expect(sample).toMatchSnapshot()
  }))

  it('with custom deep function', apply(async () => {
    const sample = await getSample('root', {
      deep: ignoreCustomDeepFunc
    })

    expect(sample).toMatchSnapshot()
  }))

  it('with custom chooser and deep function', apply(async () => {
    const sample = await getSample('root', {
      choose: chooseTypeScript,
      deep: ignoreCustomDeepFunc
    })

    expect(sample).toMatchSnapshot()
  }))
})

describe('with javascript files', () => {
  const {apply} = createVirtualEnvironment(true)

  it('without options', apply(async () => {
    const sample = await getSample('root')
    expect(sample).toMatchSnapshot()
  }))

  it('with custom chooser', apply(async () => {
    const sample = await getSample('root', {
      choose: chooseTypeScript
    })

    expect(sample).toMatchSnapshot()
  }))

  it('with custom deep function', apply(async () => {
    const sample = await getSample('root', {
      deep: ignoreCustomDeepFunc
    })

    expect(sample).toMatchSnapshot()
  }))

  it('with custom chooser and deep function', apply(async () => {
    const sample = await getSample('root', {
      choose: chooseTypeScript,
      deep: ignoreCustomDeepFunc
    })

    expect(sample).toMatchSnapshot()
  }))
})

async function getSample (dirname: string, options?: listManifestFiles.Options) {
  const sample = await listManifestFiles(dirname, options)
  return sample.map(({path, type}) => ({path, type}))
}
