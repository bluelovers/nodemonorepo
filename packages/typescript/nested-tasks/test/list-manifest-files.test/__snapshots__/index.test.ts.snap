// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`without options 1`] = `
Array [
  Object {
    "path": "root/script-only/task.js",
    "type": "module",
  },
  Object {
    "path": "root/task.js",
    "type": "module",
  },
  Object {
    "path": "root/task.yaml",
    "type": "yaml",
  },
  Object {
    "path": "root/yaml-only/task.yaml",
    "type": "yaml",
  },
]
`;
